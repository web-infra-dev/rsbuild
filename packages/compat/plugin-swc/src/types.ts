import type {
  EnvConfig,
  Extensions,
  JsMinifyOptions,
  ReactConfig,
  TransformConfig,
} from '@modern-js/swc-plugins';
import type _ from 'lodash';

export type {
  Output,
  TransformConfig,
  JsMinifyOptions,
} from '@modern-js/swc-plugins';

export type OuterExtensions = Omit<Extensions, 'ssrLoaderId' | 'configRoutes'>;

export interface ObjPluginSwcOptions<T extends 'inner' | 'outer' = 'inner'>
  extends TransformConfig {
  presetReact?: ReactConfig;
  presetEnv?: EnvConfig;

  jsMinify?: boolean | JsMinifyOptions;
  cssMinify?: boolean | CssMinifyOptions;

  /**
   * Specifies whether to modularize the import of [lodash](https://npmjs.com/package/lodash)
   * and remove unused lodash modules to reduce the code size of lodash.
   */
  transformLodash?: boolean;

  extensions?: T extends 'inner' ? Extensions : OuterExtensions;

  overrides?: Override<T>[];
}

export type Override<T extends 'inner' | 'outer'> = Omit<
  ObjPluginSwcOptions<T>,
  'test' | 'include' | 'exclude'
> & {
  test?: RegExp;
  include?: RegExp[];
  exclude?: RegExp[];
};

export type FnPluginSwcOptions = (
  config: TransformConfig,
  utilities: Utilities,
) => TransformConfig | undefined;

export type PluginSwcOptions<T extends 'inner' | 'outer' = 'inner'> =
  | ObjPluginSwcOptions<T>
  | FnPluginSwcOptions;

interface Utilities {
  mergeConfig: typeof _.merge;
  setConfig: typeof _.set;
}

export interface CssMinifyOptions {
  sourceMap?: boolean;
  inlineSourceContent?: boolean;
}
