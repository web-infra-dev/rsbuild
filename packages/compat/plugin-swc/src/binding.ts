import { type Output, Compiler as RawCompiler } from '@modern-js/swc-plugins';
import type { TransformConfig } from './types';

export {
  minify,
  minifySync,
  minifyCss,
  minifyCssSync,
} from '@modern-js/swc-plugins';

export class Compiler extends RawCompiler {
  config: TransformConfig;

  constructor(finalConfig: TransformConfig) {
    super(finalConfig);
    this.config = finalConfig;
  }
}

export function transformSync(
  config: Required<TransformConfig>,
  filename: string,
  code: string,
  map?: string,
): Output {
  const compiler = new Compiler(config);

  return compiler.transformSync(filename, code, map);
}

export function transform(
  config: Required<TransformConfig>,
  filename: string,
  code: string,
  map?: string,
): Promise<Output> {
  let compiler: Compiler;
  try {
    compiler = new Compiler(config);
  } catch (e) {
    throw new Error(`[plugin-swc] Failed to initialize config: \n${e}`);
  }
  return compiler.transform(filename, code, map);
}
