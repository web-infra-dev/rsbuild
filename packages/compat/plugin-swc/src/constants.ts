import path from 'node:path';

export const CORE_JS_PKG_PATH = require.resolve('core-js/package.json');
export const CORE_JS_DIR = path.dirname(CORE_JS_PKG_PATH);
export const SWC_HELPERS_DIR = path.dirname(
  require.resolve('@swc/helpers/package.json'),
);
export const JS_REGEX = /\.js$/;
