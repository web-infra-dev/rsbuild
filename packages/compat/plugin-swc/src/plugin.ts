import path from 'node:path';
import type { RsbuildPlugin } from '@rsbuild/core';
import { SCRIPT_REGEX, applyScriptCondition } from '@rsbuild/shared';
import { SwcMinimizerPlugin } from './minimizer';
import type {
  ObjPluginSwcOptions,
  PluginSwcOptions,
  TransformConfig,
} from './types';
import {
  applyPluginConfig,
  checkUseMinify,
  removeUselessOptions,
} from './utils';

/**
 * In this plugin, we do:
 * - Remove Babel loader if exists
 * - Add our own swc loader
 * - Remove JS minifier
 * - Add swc minifier plugin
 */
export const pluginSwc = (options: PluginSwcOptions = {}): RsbuildPlugin => ({
  name: 'rsbuild-webpack:swc',

  setup(api) {
    if (api.context.bundlerType === 'rspack') {
      return;
    }

    // first config is the main config
    let mainConfig: ObjPluginSwcOptions<'inner'>;

    api.modifyBundlerChain({
      // loader should be applied in the pre stage for customizing
      order: 'pre',
      handler: async (chain, utils) => {
        const { CHAIN_ID, environment } = utils;
        const rsbuildConfig = api.getNormalizedConfig({ environment });
        const { rootPath } = api.context;

        const { browserslist } = api.context.environments[environment];
        const swcConfigs = await applyPluginConfig(
          options,
          utils,
          rsbuildConfig,
          rootPath,
          browserslist,
        );

        // If babel plugin is used, replace babel-loader
        if (chain.module.rules.get(CHAIN_ID.RULE.JS)) {
          chain.module.rule(CHAIN_ID.RULE.JS).uses.delete(CHAIN_ID.USE.BABEL);
          chain.module.delete(CHAIN_ID.RULE.TS);
        } else {
          applyScriptCondition({
            rule: chain.module.rule(CHAIN_ID.RULE.JS),
            chain,
            config: rsbuildConfig,
            context: api.context,
            includes: [],
            excludes: [],
          });
        }

        for (let i = 0; i < swcConfigs.length; i++) {
          const { test, include, exclude, swcConfig } = swcConfigs[i];

          const ruleId =
            i > 0 ? CHAIN_ID.RULE.JS + i.toString() : CHAIN_ID.RULE.JS;
          const rule = chain.module.rule(ruleId);

          // Insert swc loader and plugin
          rule
            .test(test || SCRIPT_REGEX)
            .use(CHAIN_ID.USE.SWC)
            .loader(path.resolve(__dirname, './loader.cjs'))
            .options(removeUselessOptions(swcConfig) satisfies TransformConfig);

          if (include) {
            for (const extra of include) {
              rule.include.add(extra);
            }
          }

          if (exclude) {
            for (const extra of exclude) {
              rule.exclude.add(extra);
            }
          }
        }

        // first config is the main config
        mainConfig = swcConfigs[0].swcConfig;

        if (chain.module.rules.get(CHAIN_ID.RULE.JS_DATA_URI)) {
          chain.module
            .rule(CHAIN_ID.RULE.JS_DATA_URI)
            .uses.delete(CHAIN_ID.USE.BABEL)
            .end();
        }

        chain.module
          .rule(CHAIN_ID.RULE.JS_DATA_URI)
          .mimetype({
            or: ['text/javascript', 'application/javascript'],
          })
          // compatible with legacy packages with type="module"
          // https://github.com/webpack/webpack/issues/11467
          .resolve.set('fullySpecified', false)
          .end()
          .use(CHAIN_ID.USE.SWC)
          .loader(path.resolve(__dirname, './loader.cjs'))
          .options(removeUselessOptions(mainConfig) satisfies TransformConfig);
      },
    });

    api.modifyBundlerChain((chain, { CHAIN_ID, isProd, environment }) => {
      const rsbuildConfig = api.getNormalizedConfig({ environment });

      if (checkUseMinify(mainConfig, rsbuildConfig, isProd)) {
        const { minify } = rsbuildConfig.output;
        const minifyJs =
          minify === true || (typeof minify === 'object' && minify.js);
        const minifyCss =
          minify === true || (typeof minify === 'object' && minify.css);

        if (minifyJs) {
          chain.optimization
            .minimizer(CHAIN_ID.MINIMIZER.JS)
            .use(SwcMinimizerPlugin, [
              {
                jsMinify: mainConfig.jsMinify ?? mainConfig.jsc?.minify ?? true,
                rsbuildConfig,
              },
            ]);
        }

        if (minifyCss) {
          chain.optimization
            .minimizer(CHAIN_ID.MINIMIZER.CSS)
            .use(SwcMinimizerPlugin, [
              {
                cssMinify: minifyCss ? mainConfig.cssMinify ?? true : false,
                rsbuildConfig,
              },
            ]);
        }
      }
    });
  },
});

/// default swc configuration
export function getDefaultSwcConfig(): TransformConfig {
  const cwd = process.cwd();
  return {
    cwd,
    jsc: {
      externalHelpers: true,
      parser: {
        tsx: true,
        syntax: 'typescript',
        decorators: true,
      },
      transform: {
        react: {
          runtime: 'automatic',
        },
      },
      // Avoid the webpack magic comment to be removed
      // https://github.com/swc-project/swc/issues/6403
      preserveAllComments: true,
    },
  };
}
