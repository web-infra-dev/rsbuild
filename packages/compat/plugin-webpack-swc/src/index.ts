export { default as SwcLoader } from './loader';
export { pluginSwc } from './plugin';
export * from './binding';
export * from './types';
