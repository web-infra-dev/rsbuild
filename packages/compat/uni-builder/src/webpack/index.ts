import {
  createRsbuild,
  type RsbuildConfig,
  type RsbuildPlugin,
  type RsbuildInstance,
} from '@rsbuild/core';
import type { RsbuildTarget } from '@rsbuild/shared';
import type { UniBuilderWebpackConfig } from '../types';
import type { CreateWebpackBuilderOptions } from '../types';
import { parseCommonConfig } from '../shared/parseCommonConfig';
import { pluginModuleScopes } from './plugins/moduleScopes';
import { pluginStyledComponents } from './plugins/styledComponents';
import { pluginBabel } from './plugins/babel';
import { pluginReact } from './plugins/react';
import { withDefaultConfig } from './defaults';

export async function parseConfig(
  uniBuilderConfig: UniBuilderWebpackConfig,
  cwd: string,
  frameworkConfigPath?: string,
  target?: RsbuildTarget | RsbuildTarget[],
): Promise<{
  rsbuildConfig: RsbuildConfig;
  rsbuildPlugins: RsbuildPlugin[];
}> {
  const { rsbuildConfig, rsbuildPlugins } = await parseCommonConfig<'webpack'>(
    uniBuilderConfig,
    cwd,
    frameworkConfigPath,
    target,
  );

  rsbuildPlugins.push(
    pluginBabel({
      babelLoaderOptions: uniBuilderConfig.tools?.babel,
    }),
  );
  rsbuildPlugins.push(pluginReact());

  if (uniBuilderConfig.tools?.tsLoader) {
    const { pluginTsLoader } = await import('./plugins/tsLoader');
    rsbuildPlugins.push(
      pluginTsLoader(
        uniBuilderConfig.tools.tsLoader,
        uniBuilderConfig.tools.babel,
      ),
    );
  }

  if (uniBuilderConfig.output?.enableAssetManifest) {
    const { pluginManifest } = await import('./plugins/manifest');
    rsbuildPlugins.push(pluginManifest());
  }

  if (uniBuilderConfig.security?.sri) {
    const { pluginSRI } = await import('./plugins/sri');
    rsbuildPlugins.push(pluginSRI(uniBuilderConfig.security?.sri));
  }

  if (uniBuilderConfig.experiments?.lazyCompilation) {
    const { pluginLazyCompilation } = await import('./plugins/lazyCompilation');
    rsbuildPlugins.push(
      pluginLazyCompilation(uniBuilderConfig.experiments?.lazyCompilation),
    );
  }

  rsbuildPlugins.push(
    pluginStyledComponents(uniBuilderConfig.tools?.styledComponents),
  );

  return {
    rsbuildConfig,
    rsbuildPlugins,
  };
}

export async function createWebpackBuilder(
  options: CreateWebpackBuilderOptions,
): Promise<RsbuildInstance> {
  const { rsbuildConfig, rsbuildPlugins } = await parseConfig(
    withDefaultConfig(options.config),
    options.cwd,
    options.frameworkConfigPath,
    options.target,
  );

  const { webpackProvider } = await import('@rsbuild/webpack');
  const rsbuild = await createRsbuild({
    rsbuildConfig,
    provider: webpackProvider,
    target: options.target || 'web',
  });

  rsbuild.addPlugins([
    ...rsbuildPlugins,
    pluginModuleScopes(options.config.source?.moduleScopes),
  ]);

  return rsbuild;
}
