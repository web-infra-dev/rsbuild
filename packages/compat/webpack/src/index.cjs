/**
 * CommonJS wrapper
 */
module.exports.webpackProvider = (...args) =>
  import('./index.js').then((i) => i.webpackProvider(...args));
