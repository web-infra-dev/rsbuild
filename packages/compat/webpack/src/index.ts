export { webpackProvider } from './provider';
export type { WebpackProvider } from './provider';
export { webpackBuild } from './core/build';
export { createDefaultConfig } from './config/defaults';
export type {
  LessLoaderOptions,
  SassLoaderOptions,
  PostCSSLoaderOptions,
  CSSLoaderOptions,
  StyleLoaderOptions,
} from '@rsbuild/shared';
export type {
  RsbuildPlugin,
  RsbuildPluginAPI,

  // Config Types
  DevConfig,
  HtmlConfig,
  ToolsConfig,
  OutputConfig,
  SourceConfig,
  RsbuildConfig,
  SecurityConfig,
  NormalizedConfig,
  PerformanceConfig,
  ExperimentsConfig,

  // Hook Callback Types
  ModifyWebpackChainFn,
  ModifyWebpackConfigFn,
  ModifyWebpackChainUtils,
  ModifyWebpackConfigUtils,

  // Third Party Types
  webpack,
  WebpackChain,
  WebpackConfig,
  CSSExtractOptions,
  HTMLPluginOptions,
} from './types';
