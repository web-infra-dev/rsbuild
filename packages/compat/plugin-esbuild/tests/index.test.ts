import { createStubRsbuild } from '@rsbuild/test-helper';
import { webpackProvider } from '@rsbuild/webpack';
import { pluginEsbuild } from '../src';

describe('plugin-esbuild', () => {
  it('should set esbuild-loader', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginEsbuild()],
      rsbuildConfig: {
        provider: webpackProvider,
      },
    });
    const config = await rsbuild.unwrapConfig();

    expect(config).toMatchSnapshot();
  });

  it('should set esbuild minimizer in production', async () => {
    process.env.NODE_ENV = 'production';
    const rsbuild = await createStubRsbuild({
      plugins: [pluginEsbuild()],
      rsbuildConfig: {
        provider: webpackProvider,
      },
    });
    const config = await rsbuild.unwrapConfig();
    expect(config).toMatchSnapshot();

    process.env.NODE_ENV = 'test';
  });

  it('should not set format iife when target is not web', async () => {
    process.env.NODE_ENV = 'production';
    const rsbuild = await createStubRsbuild({
      plugins: [pluginEsbuild()],
      rsbuildConfig: {
        provider: webpackProvider,
        output: {
          targets: ['node'],
        },
      },
    });
    const config = await rsbuild.unwrapConfig();
    expect(config).toMatchSnapshot();

    process.env.NODE_ENV = 'test';
  });
});
