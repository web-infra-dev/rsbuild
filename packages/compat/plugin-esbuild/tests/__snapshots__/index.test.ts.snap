// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plugin-esbuild > should not set format iife when target is not web 1`] = `
{
  "module": {
    "rules": [
      {
        "test": /\\\\\\.\\(js\\|mjs\\|cjs\\|jsx\\)\\$/,
        "use": [
          {
            "loader": "<ROOT>/packages/compat/plugin-esbuild/compiled/esbuild-loader/index.js",
            "options": {
              "charset": "ascii",
              "loader": "jsx",
              "target": "es2015",
            },
          },
        ],
      },
      {
        "include": [
          {
            "and": [
              "<ROOT>",
              {
                "not": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
              },
            ],
          },
          /\\\\\\.\\(ts\\|tsx\\|jsx\\|mts\\|cts\\)\\$/,
        ],
        "test": /\\\\\\.\\(ts\\|mts\\|cts\\|tsx\\)\\$/,
        "use": [
          {
            "loader": "<ROOT>/packages/compat/plugin-esbuild/compiled/esbuild-loader/index.js",
            "options": {
              "charset": "ascii",
              "loader": "tsx",
              "target": "es2015",
            },
          },
        ],
      },
    ],
  },
  "optimization": {
    "minimizer": [
      ESBuildMinifyPlugin {
        "options": {
          "css": true,
          "format": undefined,
          "legalComments": "inline",
          "minify": true,
          "target": "es2015",
        },
        "transform": [Function],
      },
    ],
  },
}
`;

exports[`plugin-esbuild > should set esbuild minimizer in production 1`] = `
{
  "module": {
    "rules": [
      {
        "test": /\\\\\\.\\(js\\|mjs\\|cjs\\|jsx\\)\\$/,
        "use": [
          {
            "loader": "<ROOT>/packages/compat/plugin-esbuild/compiled/esbuild-loader/index.js",
            "options": {
              "charset": "ascii",
              "loader": "jsx",
              "target": "es2015",
            },
          },
        ],
      },
      {
        "include": [
          {
            "and": [
              "<ROOT>",
              {
                "not": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
              },
            ],
          },
          /\\\\\\.\\(ts\\|tsx\\|jsx\\|mts\\|cts\\)\\$/,
        ],
        "test": /\\\\\\.\\(ts\\|mts\\|cts\\|tsx\\)\\$/,
        "use": [
          {
            "loader": "<ROOT>/packages/compat/plugin-esbuild/compiled/esbuild-loader/index.js",
            "options": {
              "charset": "ascii",
              "loader": "tsx",
              "target": "es2015",
            },
          },
        ],
      },
    ],
  },
  "optimization": {
    "minimizer": [
      ESBuildMinifyPlugin {
        "options": {
          "css": true,
          "format": "iife",
          "legalComments": "inline",
          "minify": true,
          "target": "es2015",
        },
        "transform": [Function],
      },
    ],
  },
}
`;

exports[`plugin-esbuild > should set esbuild-loader 1`] = `
{
  "module": {
    "rules": [
      {
        "test": /\\\\\\.\\(js\\|mjs\\|cjs\\|jsx\\)\\$/,
        "use": [
          {
            "loader": "<ROOT>/packages/compat/plugin-esbuild/compiled/esbuild-loader/index.js",
            "options": {
              "charset": "ascii",
              "loader": "jsx",
              "target": "es2015",
            },
          },
        ],
      },
      {
        "include": [
          {
            "and": [
              "<ROOT>",
              {
                "not": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
              },
            ],
          },
          /\\\\\\.\\(ts\\|tsx\\|jsx\\|mts\\|cts\\)\\$/,
        ],
        "test": /\\\\\\.\\(ts\\|mts\\|cts\\|tsx\\)\\$/,
        "use": [
          {
            "loader": "<ROOT>/packages/compat/plugin-esbuild/compiled/esbuild-loader/index.js",
            "options": {
              "charset": "ascii",
              "loader": "tsx",
              "target": "es2015",
            },
          },
        ],
      },
    ],
  },
}
`;
